tests/python/test_ad_for_fwd.py

```diff
@@ -212,3 +212,142 @@ def double_for():
 
     for i in range(N):
         assert f.dual[i] == 2 * i
+
+
+@test_utils.test(exclude=[ti.cc])
+def test_double_for_loops_more_nests():
+    N = 6
+    a = ti.field(ti.f32, shape=N, needs_dual=True)
+    b = ti.field(ti.f32, shape=N, needs_dual=True)
+    c = ti.field(ti.i32, shape=(N, N // 2))
+    f = ti.field(ti.f32, shape=(N, N // 2), needs_dual=True)
+
+    @ti.kernel
+    def double_for():
+        for i in range(N):
+            for k in range(N // 2):
+                weight = 1.0
+                for j in range(c[i, k]):
+                    weight *= a[i]
+                s = 0.0
+                for j in range(c[i, k] * 2):
+                    s += weight + b[i]
+                f[i, k] = s
+
+    a.fill(2)
+    b.fill(1)
+
+    for i in range(N):
+        for k in range(N // 2):
+            c[i, k] = i + k
+
+    double_for()
+
+    for i in range(N):
+        for k in range(N // 2):
+            assert f[i, k] == 2 * (i + k) * (1 + 2**(i + k))
+
+    with ti.ad.FwdMode(loss=f, parameters=a, seed=[1.0 for _ in range(N)]):
+        double_for()
+
+    for i in range(N):
+        total_grad_a = 0
+        for k in range(N // 2):
+            total_grad_a = 2 * (i + k)**2 * 2**(i + k - 1)
+            assert f.dual[i, k] == total_grad_a
+
+    with ti.ad.FwdMode(loss=f, parameters=b, seed=[1.0 for _ in range(N)]):
+        double_for()
+
+    for i in range(N):
+        total_grad_b = 0
+        for k in range(N // 2):
+            total_grad_b = 2 * (i + k)
+            assert f.dual[i, k] == total_grad_b
+
+
+@test_utils.test(exclude=[ti.cc])
+def test_complex_body():
+    N = 5
+    a = ti.field(ti.f32, shape=N, needs_dual=True)
+    b = ti.field(ti.f32, shape=N, needs_dual=True)
+    c = ti.field(ti.i32, shape=N)
+    f = ti.field(ti.f32, shape=N, needs_dual=True)
+    g = ti.field(ti.f32, shape=N, needs_dual=False)
+
+    @ti.kernel
+    def complex():
+        for i in range(N):
+            weight = 2.0
+            tot = 0.0
+            tot_weight = 0.0
+            for j in range(c[i]):
+                tot_weight += weight + 1
+                tot += (weight + 1) * a[i]
+                weight = weight + 1
+                weight = weight * 4
```

Comment:
These might be removed by compiler? maybe use atomic add instead?

Reply:
The local atomic add will still be demoted to `load-add-store` before the autodiff pass?

Reply:
aha you're right!